crates/sui-framework/packages/sui-framework/sources/random.move

```diff
@@ -0,0 +1,128 @@
+    const ENotSystemAddress: u64 = 0;
+    const EWrongInnerVersion: u64 = 1;
+
+    const CurrentVersion: u64 = 1;
+
+    /// Singleton shared object which stores the global randomness state.
+    /// The actual state is stored in a dynamic field of type RandomnessStateInner to support
+    /// future versions of the randomness state.
+    struct Random has key {
+        id: UID,
+        version: u64,
+    }
+
+    struct RandomInner has store {
+        version: u64,
+
+        epoch: u64,
+        randomness_round: u64,
+        random_bytes: vector<u8>,
+    }
+
+    #[allow(unused_function)]
+    /// Create and share the Random object. This function is called exactly once, when
+    /// the Random object is first created.
+    /// Can only be called by genesis or change_epoch transactions.
+    fun create(ctx: &TxContext) {
+        assert!(tx_context::sender(ctx) == @0x0, ENotSystemAddress);
+
+        let version = CurrentVersion;
+
+        let inner = RandomInner {
+            version,
+            epoch: tx_context::epoch(ctx),
+            randomness_round: 0,
+            random_bytes: vector[],
+        };
+
+        let self = Random {
+            id: object::randomness_state(),
+            version,
+        };
+
+        dynamic_field::add(&mut self.id, version, inner);
+        transfer::share_object(self);
+    }
+
+    #[test_only]
+    public fun create_for_testing(ctx: &TxContext) {
+        create(ctx);
+    }
+
+    fun load_inner_mut(
+        self: &mut Random,
+    ): &mut RandomInner {
+        let version = self.version;
+
+        // replace this with a lazy update function when we add a new version of the inner object.
+        assert!(version == CurrentVersion, EWrongInnerVersion);
+
+        let inner: &mut RandomInner = dynamic_field::borrow_mut(&mut self.id, self.version);
+
+        assert!(inner.version == version, EWrongInnerVersion);
+        inner
+    }
+
+    #[allow(unused_function)] // TODO: remove annotation after implementing user-facing API
+    fun load_inner(
+        self: &Random,
+    ): &RandomInner {
+        let version = self.version;
+
+        // replace this with a lazy update function when we add a new version of the inner object.
+        assert!(version == CurrentVersion, EWrongInnerVersion);
+let inner: &RandomInner = dynamic_field::borrow(&self.id, self.version);
+        assert!(inner.version == version, EWrongInnerVersion);
+        inner
+    }
+
+    #[allow(unused_function)]
+    /// Record new randomness. Called when executing the RandomnessStateUpdate system
+    /// transaction.
+    fun update_randomness_state(
+        self: &mut Random,
+        new_round: u64,
+        new_bytes: vector<u8>,
+        ctx: &TxContext,
+    ) {
+        // Validator will make a special system call with sender set as 0x0.
+        assert!(tx_context::sender(ctx) == @0x0, ENotSystemAddress);
+
+        // Randomness should only be incremented. Ignore out-of-order updates.
```

Comment:
When will there be out-of-order updates? Or are we just trying to be cautious here?
Specifically, I don't fully understand why we need epoch and randomness_round?

Reply:
There shouldn't be out of order updates in normal operation, but there will be duplicates in normal operation. However a byzantine validator could send out of order updates if they wanted (just replaying old values). Refusing that in narhwal headers seemed too punitive to validators that may just be a bit behind, it's easy enough to ignore.

Reply:
> a byzantine validator could send out of order updates if they wanted

How does this work? I would assume that everyone will need to agree on every transaction which cannot be controlled by a single validator.

Reply:
They would just replay old values of randomness

Reply:
To elaborate, any validator can include a full randomness signature in its narwhal header which will be accepted if the sig is valid

Reply:
Discussed this further with Mark, actually we think existing duplicate consensus tx supression should prevent replays. Changed this to an assert